Initial work to add a Time type to Base.Dates

quinnj opened the pull request:
Addresses #12140.

TODO:
- [x] Figure out the best way to "display" (i.e. `show`) micro/nanoseconds
- [x] Address any other needed functionality: @aviks @tbreloff @ScottPJones, anybody else interested
- [x] Add lots of tests

quinnj commented:
[CI cancelled because it's not worth running at this point]

ivarne commented on base/dates/time.jl:
```diff
@@ -0,0 +1,58 @@
+(+)(x::Time,y::Dates.TimePeriod)   = return Time(NS(value(x)+Dates.tons(y)))
+(-)(x::Time,y::Dates.TimePeriod)   = return Time(NS(value(x)-Dates.tons(y)))
+(+)(y::Dates.TimePeriod,x::Time) = x + y
+(-)(y::Dates.TimePeriod,x::Time) = x - y
```
Do you intend `2 hours - 12:30` to be `10:30`?

ivarne commented on base/dates/time.jl:
```diff
@@ -0,0 +1,58 @@
+tons(x::Nanosecond) = x
+
+(+)(x::Time,y::Dates.TimePeriod)   = return Time(NS(value(x)+Dates.tons(y)))
+(-)(x::Time,y::Dates.TimePeriod)   = return Time(NS(value(x)-Dates.tons(y)))
```
I don't see why we want a bounds check in the constructor, but as long as we do, we should have it in the arithmetic too.

ivarne commented on base/dates/time.jl:
```diff
@@ -0,0 +1,58 @@
+    micros::Microsecond=Microsecond(0),
+    nanos::Nanosecond=Nanosecond(0)) = Time(Int64(h),Int64(m),Int64(s),Int64(millis),Int64(micros),Int64(nanos))
+_c(c) = convert(Int64,c)
+Time(h,m=0,s=0,millis=0,micros=0,nanos=0) = Time(_c(h),_c(m),_c(s),_c(millis),_c(micros),_c(nanos))
```
Am I right that this will match `Time(Minute(12))` and result in 12:00? Maybe restrict to `:> Real`?

nalimilan replied:
And what's the difference between `_c(c) = convert(Int64,c)` and `Int64()`?

tbreloff commented:
Jacob: thanks a `tons` for this.  I certainly think you're a big chunk of the way there.  I have a few questions/comments:
- Is there a reason the `Time` type isn't a subtype of `TimePeriod` with the same definition as `Nanosecond`?  If this type is intended to fill the same role as my `TimeOfDay` type, then I would think this is just a specialized `TimePeriod` that happens to convert numerically the same way as a `Nanosecond`.
- Can you define ranges of `Time` types?  
- I'm undecided on the checks in the constructor... as @ivarne has pointed out, if you have them in the constructor, they should be in all mutating operations as well, and the efficiency of that worries me.  Maybe there could be a specialized constructor that does bounds checking for people that need it?
- Should line 7 also be named `tons` instead of `NS`? Can `tons` be a different name?  I can't stop reading it as one word.

Regarding printing/parsing... I would personally like to see methods like:

```
function TimeOfDay(str::String)
  tmp = split(str,":")
  minutes = 0
  seconds = 0
  nanos = 0
  
  hours = parse(Int, tmp[1])
  if length(tmp) > 1
    minutes = parse(Int, tmp[2])
  end

  if length(tmp) > 2
    tmp2 = split(tmp[3], ".")
    seconds = parse(Int, tmp2[1])
    if length(tmp2) > 1
      nanos = parse(Int, rpad(tmp2[2],9,"0")[1:9])
    end
  end

  TimeOfDay(hours * nanosInOneHour + minutes * nanosInOneMinute + seconds * nanosInOneSecond + nanos)
end

function Base.string(timeOfDay::TimeOfDay)
  secsSinceMidnight, nanos = divrem(timeOfDay.nanosSinceMidnight, nanosInOneSecond)
  hours, hourrem = divrem(secsSinceMidnight, secondsInOneHour)
  minutes, seconds = divrem(hourrem, secondsInOneMinute)
  microseconds = div(nanos, millisInOneSecond)
  string(lpad(hours,2,"0"), ":", lpad(minutes,2,"0"), ":", lpad(seconds,2,"0"), ".", lpad(microseconds,6,"0"))
end
```

Your string/print is very similar, but I don't think I saw a string-based constructor.  Of course, once this becomes part of Base, we could specialize some formatting options in the Formatting.jl package.

Thanks again for the effort!

JeffreySarnoff commented:
postfixes for subsecond units

Unit of Time | Postfix
------------ | -------------
seconds | s
milliseconds |  ms
microseconds |  us
nanoseconds |  ns
picoseconds |  ps
femtoseconds | fs
attoseconds | as
zeptoseconds | zs
yoctoseconds | ys

```
100 seconds:      100 s   ( "100s"  )
100 microseconds: 100 us  ( "100us" )
100 nanoseconds:  100 ns  ( "100ns" ) 
```

JeffreySarnoff commented:
As time of day is frequently 'flattened' into seconds or subseconds of a day and the preferred subsecond resolution varies with the application aor the computational environment, one approach to processing a time of day to allow parametric content or to offer a defaulted argument or to let the content of the string or of the tuple  (and better, a prioritized mix of, say, two of those) to select working subsecond resolution.

ViralBShah commented:
Bump. Would be nice to get this one done.

JeffreySarnoff commented:
Just a few notes and a question:

Time must use an integer underlying type, floats don't work well
because the spacing between adjacent values is not constant.

A week is the largest amount of time in common time units that is 
a constant number of seconds (604_800s). If we want to represent
weeks in seconds then Int64 allows picosecond (1e-12) resolution
(1/10 picosecond if that helps).  If we let go of weeks as a 
quantity of seconds, then days become the largest amount of time
in common time units (86_400s). Then, Int64 allows picosecond
(1/100 picosecond if that helps) resolution.  To cover smaller
time intervals requires Int128 -- which covers, for example
1/10 attosecond through 10 exaseconds.  Int128 allows encoding
the age of the universe in attoseconds.

Given the state of the art in ultra-high frequency physics, where
femtosecond timing is in use and attosecond timing is useful, imo
limiting our internal resolution to picoseconds would be an error
because it means having a second high-resolution timing library.
Int128 runs quickly; so performance should not be a consideration.

What is of greater import, keeping the storage size to 64 bits
or having a single sub-second time type that covers all spans?

tkelman commented:
Int128 is not very well-supported on all platforms or compilers.

JeffreySarnoff commented:
that is a shame, and a deciding factor.

On Tue, Nov 24, 2015 at 1:58 PM, Tony Kelman <notifications@github.com>
wrote:

> Int128 is not very well-supported on all platforms or compilers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/12274#issuecomment-159372587>.
>

JeffreySarnoff commented:
I am not the guy to fold Time in with Date; however if something along
these lines would be helpful, I could make it. And if not, that's perfectly
ok:

#=
    1e18 yoctoseconds per microsecond
    86_400*1e6 microseconds/day
    +/- 200_000 years in microseconds
    microsecond #0 is at POSIX time 0 (1970-01-01 00:00:00 GMT)

    To simplify this sketch by avoiding a great deal
    of bounds checking, we adopt the convention that
    times specified in, and times requested in
    sub-microsecond units have no microsecond component (x.u==0).

=#

immutable RelativeTime
    u::Int64 # MicroSeconds, 0..±1e18-1
    y::Int64 # YoctoSeconds, 0..1e18-1, y >= 0
end

immutable GivenTime
    u::Int64 # MicroSeconds, 0..±1e18-1, 0 is at POSIX time 0
    y::Int64 # YoctoSeconds, 0..1e18-1, y >= 0
end

Times = Union{RelativeTime, GivenTime}
N64   = Union{Int64, Float64}
idiv{T<:N64,U<:N64}(p::T,q::U) = round(Int64, div(p,q))

Yoctoseconds{T<:Times}(t::T) = t.y
Zeptoseconds{T<:Times}(t::T) = div(t.y, Int64(1e3))
Attoseconds{T<:Times}(t::T)  = div(t.y, Int64(1e6))
Femtoseconds{T<:Times}(t::T) = div(t.y, Int64(1e9))
Picoseconds{T<:Times}(t::T)  = div(t.y, Int64(1e12))
Nanoseconds{T<:Times}(t::T)  = div(t.y, Int64(1e15))
Microseconds{T<:Times}(t::T) = t.u
Milliseconds{T<:Times}(t::T) = div(t.u,        Int64(1e3))
Seconds{T<:Times}(t::T)      = div(t.u,        Int64(1e6))
Minutes{T<:Times}(t::T)      = div(t.u,     60*Int64(1e6))
Hours{T<:Times}(t::T)        = div(t.u,   3600*Int64(1e6))
Days{T<:Times}(t::T)         = div(t.u,  86400*Int64(1e6))
Weeks{T<:Times}(t::T)        = div(t.u, 604800*Int64(1e6))

function dSeconds(t::Float64)
   isneg, abst = signbit,abs(t)
   ...fldmod..
end
dSeconds(t::T)  = RelativeTime(idiv(t,Int64(1e6)),0)
dMicroseconds{T<:N64}(t::T) = RelativeTime(t,0)
dYoctoseconds{T<:N64}(t::T) = RelativeTime(0,t)

nSeconds(t::T)  = GivenTime(idiv(t,Int64(1e6)),0)
nMicroseconds{T<:N64}(t::T) = GivenTime(t,0)
nYoctoseconds{T<:N64}(t::T) = GivenTime(0,t)



On Tue, Nov 24, 2015 at 2:09 PM, Jeffrey Sarnoff <jeffrey.sarnoff@gmail.com>
wrote:

> that is a shame, and a deciding factor.
>
> On Tue, Nov 24, 2015 at 1:58 PM, Tony Kelman <notifications@github.com>
> wrote:
>
>> Int128 is not very well-supported on all platforms or compilers.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/JuliaLang/julia/pull/12274#issuecomment-159372587>.
>>
>
>

tkelman commented:
What's the plan here? This has been pretty quiet and it probably isn't 0.5 material unless it gets finished up very soon.

Jeffrey-Sarnoff commented:
I took the non-response as preference for another approach -- I still am not the guy to merge Time into Date (I have not used Date and defer to @quinnj  on its construction and what best works to interpose Time within that extensive software subfacility).  OTOH, if we want to -- I'd spend the weekend doing some coding to have a Type of Time  that gives up weeks,days,hours,minutes,seconds,microseconds,milliseconds,nanoseconds maybe attoseconds if asked.  If someone else would write how to use that with Date stuff.

Jeffrey-Sarnoff commented:
@tkelman Is Int128 better supported with v0.5?

iamed2 commented:
I think there's a couple different ideas here:
1) A type representing time of day https://github.com/JuliaLang/julia/issues/12140
2) New sub-millisecond Period subtypes

Those might each be better as separate PRs. It's not really clear which one each interested party is after. 1) is probably easy to implement using the work here and in the issue. 2) might need more consideration.

quinnj commented:
I think we should focus on use-case 1). The second case is much better served by going the SIUnits.jl route, where you already get a parametric `Second` type defined for all the granularities.

I'm happy to push this through in the next few days if people really want it. I also think this would fit well in a package; are there strong feelings one way or the other at this point? (package vs. in Base)

Jeffrey-Sarnoff commented:
Hi  Jacob,
For myself, no preference.  For Julia, I prefer Time be with Date.  Perhaps going about it the way you did with Date makes sense (first a package, some shakedown and requests/remodels, then in with Base)  [why mess with a winning approach].

tkelman commented:
I think the performance issues that are just now being identified by Simon and others in the Dates code are indicative of just how long it can take to get these things right, and we probably shouldn't rush new features if you can see a good way to separate concerns with Dates in Base and Time types in a package. Some of the Dates internals might also be up for some refactoring to address those performance issues, which seem to have more people working on them right now with a higher priority, so I'd focus there at the moment.

tkelman commented:
@Jeffrey-Sarnoff Int128 probably works a bit better with the more recent LLVM versions used on master, but it could always use more testing. If we ever look more seriously at supporting a Visual-Studio built version of Julia, there may still be gaps in Int128 support there.

hgeorgako commented:
Don't know if this is the right issue to comment on this, but from a quant/trader's perspective, having the ability to merge high frequency tick data with nanosecond precision time-stamps is becoming a big deal in the industry.  R's xts library does not currently support this. I believe pandas does.  If for one would love to see this functionality enabled sooner than later.

quinnj commented:
Ok, made some progress here and this is actually getting close in terms of the rest of the infrastructure we have in Dates. (plus lots of tests, yay!)

Would definitely love some review/feedback.

One of the biggest design questions I ran into was the following:

* Do we treat `Time` more like `Date` or `DateTime` or more like a special case of a `CompoundPeriod` type? (since `Time` is essentially a more efficient representation of a CompoundPeriod composed of Hours, Minutes, Seconds, etc.)
* This most readily affects arithmetic: do we allow `x::Time + y::Time`? In the `TimeType` respect, we don't allow `Date + Date` (it doesn't' really make sense semantically), but it certainly feels a little more natural with something like `Time` (if you view it as a special-case CompoundPeriod).
* It gets a little weird though, since `Time` has somewhat of a finite range (~2.5 days using nanosecond precision); currently things would just keep wrapping around (i.e. 23 + 1 hour == Hour(0)), so it doesn't necessarily follow CompoundPeriod behavior

I think I'm leaning towards treating `Time` as a true `TimeType`, which would disallow `x::Time + y::Time` and `x::DateTime + y::Time`. You'd really want to view `Time` then as a true marker of a specific moment in a day between midnight and 11:59:59.

omus commented:
I think `Time` should be treated more like a `Date` but should also be able to be explicitly converted to a `CompoundPeriod` via `CompoundPeriod(::Time)`. I think a `Time` type should be bounded between 00:00:00 and 23:59:59.

Are we going to support `x::Time + y::Period`? Would `Time("23:00:00") + Hour(2)` roll over to be `Time("01:00:00")`?

omus commented:
Something I didn't see in the PR is `x::Date + y::Time = z::DateTime`. Maybe that isn't possible due to the extra precision in `Time`.

omus commented on base/dates/types.jl:
```diff
@@ -161,6 +182,31 @@ function Date(periods::Period...)
+Construct a `Time` type by `Period` type parts. Arguments may be in any order. `Time` parts
+not provided will default to the value of `Dates.default(period)`.
+"""
+function Time(periods::Period...)
```
Does `periods::TimePeriod...` work here?

quinnj replied:
Good catch. I'll change that.

omus commented on test/dates/accessors.jl:
```diff
@@ -119,6 +119,47 @@ function test_dates2(from,to)
+    h = mi = s = 0
+    for h = 0:23
+        for mi = 0:59
+            for s = 0:59
```
Good use case for combining nested for loops: 
```julia
for h = 0:23, mi = 0:59, s = 0:59
    ...
end
```

iamed2 commented:
> Something I didn't see in the PR is `x::Date + y::Time = z::DateTime`. Maybe that isn't possible due to the extra precision in `Time`.
It could throw an `InexactError` when not convertible, like `convert(::Type{Float64}, ::Int64)`

quinnj commented:
Ok, made some additional improvements here:

* Cleaned up `show` for Time a little (there were some corner cases)
* Defined a `CompoundPeriod(t::Time)` method to turn a `Time` into a `CompoundPeriod`
* Removed the `t::Time + r::Time` method and replaced it with a `dt::Date + t::Time => DateTime` method

This also passed all tests for me locally. I think the docs should be in pretty good shape due to @MichaelHatherly's latest port of inline dates docs.

Do give any more feedback!

omus commented on base/dates/ranges.jl:
```diff
@@ -4,10 +4,12 @@
 # Given a start and end date, how many steps/periods are in between
 guess(a::DateTime,b::DateTime,c) = floor(Int64,(Int128(b) - Int128(a))/toms(c))
 guess(a::Date,b::Date,c) = Int64(div(Int64(b - a),days(c)))
+len(a::Time,b::Time,c) = Int64(div(Int64(b - a),tons(c)))
```
Does this still work? I don't think `b::Time - a::Time` is defined.

quinnj replied:
It is, just like for other TimeTypes. We indeed want to allow seeing the difference between two TimeTypes, but their addition is was doesn't make sense semantically.

omus replied:
Ah, makes sense

omus commented on base/dates/types.jl:
```diff
@@ -58,10 +58,11 @@ immutable Date <: TimeType
+# Time is a nanosecond precision representation of hours:minutes:seconds.fractional
+immutable Time <: TimeType
+    instant::Nanosecond
+    Time(instant::Nanosecond) = new(instant)
```
Constructor is unnecessary here

omus replied:
Hmm, couldn't someone create a `Time` with a negative nanoseconds or greater than or equal to 24:00:00? Do we need to enforce that here?

quinnj replied:
Constructor is necessary, just like for the other TimeTypes. By default, every type is given two constructors, one with the exact arguments of its fields and another that allows "Any" with calls to convert of the field types. By defining just the one constructor according to the field type, we're *not* getting the "Any" constructor, which is desirable since we define our own later on (without this, we would get a method overwritten warning when we define our own "Any" fallback constructor)

omus commented on base/dates/types.jl:
```diff
@@ -169,21 +215,32 @@ calendar(dt::Date) = ISOCalendar
-Base.isless(x::Date,y::Date) = isless(value(x),value(y))
 Base.isless(x::DateTime,y::DateTime) = isless(value(x),value(y))
+Base.isless(x::Date,y::Date) = isless(value(x),value(y))
+Base.isless(x::Time,y::Time) = isless(value(x),value(y))
```
Consider doing:
```julia
Base.isless{T<:TimeType}(x::T, y::T) = isless(value(x), value(y))
```

omus commented on base/dates/types.jl:
```diff
@@ -169,21 +215,32 @@ calendar(dt::Date) = ISOCalendar
+    return hour(a) == hour(b) && minute(a) == minute(b) &&
+        second(a) == second(b) && millisecond(a) == millisecond(b) &&
+        microsecond(a) == microsecond(b) && nanosecond(a) == nanosecond(b)
+end
```
I feel like I'm missing something. Doesn't `value(a) == value(b)` suffice?

quinnj replied:
The reasoning here actually corresponds to your other comment above about allowing negative/24 hour+ values internally. Right now I'm allowing it, because the way the accessors are written, everything has a "wrapping" behavior; i.e. `Time(23)  + Hour(1) == Time(0)`. Now, internally, we can let the # of nanoseconds increase because the accessor is still correctly returning 0 for the hour. The only way this comes out is in equality; we can't simply compare the inner values since I might have two different # of nanoseconds referring to `Time(0)`. It may indeed be simpler/better to keep things clean internally, but so far it seems like it's not a big deal.

omus replied:
Ok, I was thinking we had a restricted range of allowed nanoseconds for the internal storage. The one problem I can see with this is dealing with Integer overflow:

```julia
julia> x = Time(Nanosecond(typemax(Int64)))
23:47:16.854775807

julia> x + Nanosecond(1)
00:12:43.145224192
```

quinnj replied:
Yeah, I think we just need to make a decision; do we want:

1. Simpler construction/arithmetic and automatic wrapping, but at the cost of overflow (not unlike Julia's Integer types)
2. More complex construction/arithmetic implementations, manually implemented wrapping, but no overflow

quinnj commented on test/dates/periods.jl:
```diff
@@ -292,6 +322,15 @@ dt = Dates.DateTime(2014)
+@test Dates.tons(us) == Dates.value(Dates.Nanosecond(us)) == 1000
+@test Dates.tons(ms) == Dates.value(Dates.Nanosecond(ms)) == 1000000
+@test Dates.tons(s)  == Dates.value(Dates.Nanosecond(s))  == 1000000000
+@test Dates.tons(mi) == Dates.value(Dates.Nanosecond(mi)) == Int64(60000000000)
```
@tkelman, any pointers here? 32-bit keeps failing on this test. Is it because we're unable to get the right test value? Anything else immediately obvious? Or maybe I need to go back and look at the `tons` method....

tkelman replied:
I'm not at all familiar with this code, not sure I could be of much help. I don't know whether Gallium works on 32 bit Linux, but you could try the poor-man's debugger of following the dispatch trail with `@which`. CI is still running here, but I assume the issue is happening on both 32 bit Linux and 32 bit Windows?

quinnj replied:
Yes, happening on both. I'll see if I can't get a 32-bit build on my windows machine to test it out.

tkelman replied:
This doesn't touch the C code, so you should be able to use a nightly and build_sysimg.jl - assuming that script is working at the moment.

omus replied:
@quinnj I think I found the issue `N = 1` should be `N = Int64(1)`. See https://github.com/JuliaLang/julia/pull/12274/files#diff-18ff5efda613bb05a8b004a550205c03R402

quinnj replied:
Thanks @omus, let's see how CI does now.

tkelman commented on base/dates/periods.jl:
```diff
@@ -51,7 +52,6 @@ Base.typemax{P<:Period}(::Type{P}) = P(typemax(Int64))
 # Default values (as used by TimeTypes)
 """
     default(p::Period) -> Period
-
```
why is this line being deleted?

tkelman commented:
This could use squashing, probably before merging, as the internal merge commits and commits that mostly revert one another are a bit noisy to review here.

tkelman commented on base/dates/periods.jl:
```diff
@@ -146,7 +159,6 @@ coarserperiod(::Type{Month}) = (Year,12)
 # and convert more-precise periods to less-precise periods when possible
 """
     CompoundPeriod
-
```
same here and below, I think the docstring convention has one blank line between the signature and the body

quinnj replied:
Thanks for catching. I think I've managed to clean up all the doc changes.

tkelman commented on test/dates/accessors.jl:
```diff
@@ -119,6 +119,33 @@ function test_dates2(from,to)
+        @test s == Dates.second(t)
+    end
+end
+test_times1()
```
does this get called again, or should it just be a let block?

quinnj replied:
I just followed the same pattern that was used for testing other accessors.

quinnj commented:
I'll leave this open for a few more days to get more review, but all CI checks have now passed.

tkelman commented:
docs should probably be populated in the rst, otherwise lgtm

omus commented on base/dates/adjusters.jl:
```diff
@@ -167,13 +174,38 @@ provided `y, m` arguments, and will be adjusted until `f::Function` returns `tru
+end
+function Time(func::Function,h,mi,s,ms,us;step::Dates.Period=Dates.Nanosecond(1),negate::Bool=false,limit::Int=10000)
+    return Dates.adjust(Dates.DateFunction(func,negate,Dates.Time(h,mi,s,ms,us)),Dates.Time(h,mi,s,ms,us),step,limit)
+end
```
Should be able to drop `Dates.` when working inside the `Dates` module.

omus commented on base/dates/conversions.jl:
```diff
@@ -19,8 +19,17 @@ the new `DateTime` are assumed to be zero.
+
 Base.convert(::Type{DateTime},dt::Date) = DateTime(UTM(value(dt)*86400000))
 Base.convert(::Type{Date},dt::DateTime) = Date(UTD(days(dt)))
+Base.convert(::Type{Time},dt::DateTime) = Time(Nanosecond((Dates.value(dt) % 86400000) * 1000000))
```
Could just be `value` instead of `Dates.value`

hgeorgako commented:
Is the plan to include this functionality in the initial 0.5 release?

omus commented:
@quinnj probably should be merged prior to a 0.5 RC.

quinnj commented:
I actually think I'd like to push this to 0.6. I'd like to refactor a little bit to make the internal precision parameterised on a period (i.e. you could have a Second-precision Time, Millisecond-precision, etc.).

hgeorgako commented:
bump.  Read Stefan's comment about feature freeze of 0.6 by end of year https://groups.google.com/forum/#!topic/julia-dev/tr-8URYXzfI. Would be nice to have this functionality available.

hgeorgako commented:
Can this be included by the freeze? Would be very useful to have.

StefanKarpinski commented:
It would be lovely to have this in 0.6. Any chance, @quinnj?

quinnj commented:
Sure, let me take a crack at it and try to get things cleaned up.

StefanKarpinski commented:
Any progress?

quinnj commented:
Testing locally.

quinnj committed:
Add a Time type to the Base.Dates module

quinnj commented:
Alright, tests are passing locally. A general ping to anyone who'd like to review, with a note that this has already gone through some review and very much follows the current Date/DateTime implementation, which hopefully makes it decent. I think it'd be helpful to really think if there are any pieces of functionality that aren't here yet, so speak up if this can't do something you'd expect.

cc: @omus @shashi

JeffreySarnoff commented:
I don't know why the 32bit arch fails the tests.

It would be very helpful for the distinguished units of time to work with themselves:

```julia
    > ten_millisecs = Millisecond(10)
    10 milliseconds
    > millisecond( ten_millisecs )
    # ERROR: MethodError: no method matching millisecond(::Dates.Millisecond)
    # Closest candidates are:
    #   millisecond(::Dates.Time) at /home/jas/.julia/v0.6/Dates/src/accessors.jl:59
    #   millisecond(::Dates.DateTime) at /home/jas/.julia/v0.6/Dates/src/accessors.jl:55
```

It would be very helpful for the distinguished units of time to work with each other:

  ```julia
    > ten_millisecs = Millisecond(10)
    10 milliseconds
    > five_microsecs = Microsecond(5)
    5 microseconds

    > millisecs_5000 = millisecond(five_microsecs)
    #  ERROR: MethodError: no method matching millisecond(::Dates.Microsecond)
    > millisecs_5010 = (five_microsecs + ten_millisecs)
   10 milliseconds, 5 microseconds
    > milliseconds( millisecs_5010 )
   # ERROR: MethodError: no method matching millisecond(::Dates.CompoundPeriod)
```

For conversion that overflow ( nanosecond( Day(106752) ), throw the exception.

tkelman commented:
causes a segfault on 32 bit, both windows and linux so it's not the existing issue

quinnj commented:
I'm not used to all the new testset output; is there any way to tell where the segfault happened? What was the last set of tests run? I'm a little overwhelmed with the sheer output.

tkelman commented:
scroll up - dates/accessors

omus commented on base/dates/periods.jl:
```diff
@@ -16,23 +16,24 @@ for period in (:Year, :Month, :Week, :Day, :Hour, :Minute, :Second, :Millisecond
-        Construct a `$($period_str)` object with the given `v` value. Input must be
-        losslessly convertible to an `Int64`.
-        """ $period(v)
+    typs = period in (:Microsecond, :Nanosecond) ? ["Time"] : (period in (:Hour, :Minute, :Second, :Millisecond) ? ["DateTime","Time"] : ["TimeType"])
```
Maybe this should be renamed to `types`?

omus commented on base/dates/periods.jl:
```diff
@@ -16,23 +16,24 @@ for period in (:Year, :Month, :Week, :Day, :Hour, :Minute, :Second, :Millisecond
+            Construct a `$($period_str)` object with the given `v` value. Input must be
+            losslessly convertible to an `Int64`.
+            """ $period(v)
+        end
```
Docstrings seem a little strange here as the docstring specifically for `Hour(::DateTime)` references `Time`.

```
Hour(dt::DateTime) -> Hour

The hour part of a Time or DateTime as a Hour.

Hour(dt::Time) -> Hour

The hour part of a Time or DateTime as a Hour.
```

quinnj replied:
I think I've got them better sorted out now.

omus commented on base/dates/periods.jl:
```diff
@@ -440,7 +457,10 @@ toms(c::Day)         = 86400000*value(c)
 toms(c::Month)       = 86400000.0*30.436875*value(c)
 toms(c::Year)        = 86400000.0*365.2425*value(c)
-toms(c::CompoundPeriod) = isempty(c.periods)?0.0 : Float64(sum(toms,c.periods))
+toms(c::CompoundPeriod) = isempty(c.periods)? 0.0 : Float64(sum(toms, c.periods))
```
Doesn't work when `CompoundPeriod` contains a `Microsecond` or a `Nanosecond`.
```julia
julia> Dates.toms(Dates.CompoundPeriod(Dates.Nanosecond(1)))
ERROR: MethodError: no method matching toms(::Base.Dates.Nanosecond)
Closest candidates are:
  toms(::Base.Dates.CompoundPeriod) at dates/periods.jl:460
  toms(::Base.Dates.Year) at dates/periods.jl:459
  toms(::Base.Dates.Month) at dates/periods.jl:458
  ...
```

omus commented on test/dates/types.jl:
```diff
@@ -168,6 +192,12 @@ ms = Dates.Millisecond(1)
+@test Dates.Time(h,mi,s,ms) == Dates.Time(1,1,1,1)
+@test Dates.Time(h,mi,s,ms,us) == Dates.Time(1,1,1,1,1)
+@test Dates.Time(h,mi,s,ms,us,ns) == Dates.Time(1,1,1,1,1,1)
+@test Dates.Time(us,h,s,ns,mi,ms) == Dates.Time(1,1,1,1,1,1)
```
Wouldn't these tests be better if we used different values for the various periods?

quinnj replied:
Maybe; but we check different period values in `dates/accessors.jl`, so I don't think it's a big deal.

tkelman replied:
> we check different period values in `dates/accessors.jl`

Where? Using the methods that construct a Time from mixed period inputs? Using all ones here could easily hide bugs in that constructor method.

tkelman replied:
Was not addressed. These tests would really easily pass even in the presence of major bugs in the implementation.

quinnj replied:
As I mentioned, there are actually plenty of other tests that cover different inputs, such [here](https://github.com/JuliaLang/julia/blob/f263b4ae79aa2424df6aa7f50f6af058484c37e5/test/dates/types.jl#L54), and [here](https://github.com/JuliaLang/julia/blob/f263b4ae79aa2424df6aa7f50f6af058484c37e5/test/dates/types.jl#L112), as well as [here](https://github.com/JuliaLang/julia/blob/f263b4ae79aa2424df6aa7f50f6af058484c37e5/test/dates/accessors.jl#L113).

tkelman replied:
You're missing the point of what's lacking about this. The first two places you linked to have the same basic issue of using zeros for all but one input. The third doesn't test this direction of taking time typed objects as inputs to this constructor.

quinnj commented on base/dates/arithmetic.jl:
```diff
@@ -57,14 +72,16 @@ function (-)(dt::Date,z::Month)
+(+)(x::DateTime, y::Period)   = return DateTime(UTM(value(x) + toms(y)))
+(-)(x::DateTime, y::Period)   = return DateTime(UTM(value(x) - toms(y)))
+(+)(x::Time, y::TimePeriod)   = return Time(Nanosecond(value(x) + tons(y)))
+(-)(x::Time, y::TimePeriod)   = return Time(Nanosecond(value(x) - tons(y)))
```
Better style, more legible.

omus replied:
I'll elaborate as the inline example isn't showing the surrounding context. Maybe we should remove the spaces before `=` on some of these methods? That way they would line up

```julia
(+)(x::Date, y::Week) = return Date(UTD(value(x) + 7*value(y)))
(-)(x::Date, y::Week) = return Date(UTD(value(x) - 7*value(y)))
(+)(x::Date, y::Day)  = return Date(UTD(value(x) + value(y)))
(-)(x::Date, y::Day)  = return Date(UTD(value(x) - value(y)))
(+)(x::DateTime, y::Period)   = return DateTime(UTM(value(x) + toms(y)))
(-)(x::DateTime, y::Period)   = return DateTime(UTM(value(x) - toms(y)))
(+)(x::Time, y::TimePeriod)   = return Time(Nanosecond(value(x) + tons(y)))
(-)(x::Time, y::TimePeriod)   = return Time(Nanosecond(value(x) - tons(y)))
(+)(y::Period, x::TimeType) = x + y
(-)(y::Period, x::TimeType) = x - y
```

vs.
```julia
(+)(x::Date, y::Week) = return Date(UTD(value(x) + 7*value(y)))
(-)(x::Date, y::Week) = return Date(UTD(value(x) - 7*value(y)))
(+)(x::Date, y::Day)  = return Date(UTD(value(x) + value(y)))
(-)(x::Date, y::Day)  = return Date(UTD(value(x) - value(y)))
(+)(x::DateTime, y::Period) = return DateTime(UTM(value(x) + toms(y)))
(-)(x::DateTime, y::Period) = return DateTime(UTM(value(x) - toms(y)))
(+)(x::Time, y::TimePeriod) = return Time(Nanosecond(value(x) + tons(y)))
(-)(x::Time, y::TimePeriod) = return Time(Nanosecond(value(x) - tons(y)))
(+)(y::Period, x::TimeType) = x + y
(-)(y::Period, x::TimeType) = x - y
```

quinnj committed:
Time cleanup

quinnj commented:
Weird; a 32-bit GC error is happening in `dates/accessors.jl`. I'd appreciate some help if anyone knows what's going on here; I'm not sure how a couple `fld` and `mod` calls could result in a GC segfault?

omus commented on test/dates/accessors.jl:
```diff
@@ -118,7 +109,20 @@ function test_dates2(from,to)
+# test hour, minute, second
+let h=0, mi=0, s=0, ms=0, us=0, ns=0
+    for h = 0:23, mi = 0:59, s = 0:59,
+        ms in (0,1,500,999), us in (0,1,500,999), ns in (0,1,500,999)
```
@quinnj you can mitigate the segfault by reducing how many iterations this loop has. For example changing `s = 0:59` to `s = (0,59)` seems to avoid the segfault on 32-bit.

tkelman replied:
that seems like a deeper bug...

omus replied:
Agreed. I'll take another look into this.

JeffreySarnoff replied:
Is there any more understanding of the cause?

omus replied:
The segfault seems unrelated to this PR. I'm attempting a bisect to identify the root cause.

tkelman replied:
what's the test case look like, just for reference?

omus replied:
The [new accessor test case](https://github.com/JuliaLang/julia/pull/12274/files#diff-2f3dfed48f169e03378a18a8e7db28d4R114) triggers the GC segfault. From what I've seen so far it seems that 32-bit systems have this issue when they perform a large amount of tests. Since this particular test generates 33177600 `@test` calls it appears to regularly trigger the problem.

JeffreySarnoff replied:
Thank you for the info.  It will be v,good solution --the 64 bit GC wants the same adjustment.

Knowing that nothing in this PR gives rise to the GC failure by its manner of work, rather the numerosity of that test has brought to our awareness that *deeper bug*.  Do we feel it helpful to allow this (with the  lesserly agglomerative test replacing the [new accessor test](https://github.com/JuliaLang/julia/pull/12274/files#diff-2f3dfed48f169e03378a18a8e7db28d4R114)) to merge.  I think it important that people have some time with this *Time* before v0.6 rolls into home and hearth.     

    

Any few brief wants of the dev community would make it all the nicer for the rest of us>

quinnj replied:
@shashi, @omus, I'd say let's go ahead and merge the date-parsing PR first (I think there's just a tweak or two left there), and then I'll do a rebase here and we can merge. @omus, if we can put together a minimum repro of the 32-bit segfault, we can open a separate issue with that and reduce the test case here so it doesn't fail.

omus replied:
Finally created the issue: #20107

tkelman commented:
I think this is past the deadline for inclusion in 0.6, and should be a package in the short-term.

quinnj commented:
I don't think there's anything left to do here though; I was just waiting on the datetime-parsing PR to merge before rebasing and merging this. I can rebase and merge in the next few hours.

quinnj committed:
Fix Time doc

tkelman commented on base/dates/adjusters.jl:
```diff
@@ -184,6 +191,31 @@ function DateTime(func::Function, y, m, d, h, mi, s; step::Period=Millisecond(1)
 end
 
+"""
+    Time(f::Function, h[, mi, s, ms, us]; step=Second(1), negate=false, limit=10000) -> Time
```
do you want these docstrings to be in the stdlib docs or repl-only?

tkelman replied:
the default step isn't always one second in the implementations below

quinnj replied:
Should the docstrings be in stdlib or repl-only? I don't have a preference either way, but if there's a convention we **should** follow, happy to do it. Yeah, I would hope it would be obvious that if you're doing sub-Second Time ranges, it wouldn't default to a Second (since that would potentially lose precision in range steps).

tkelman replied:
If the Dates section of the stdlib docs includes other docstrings for non-exported types, then this should be added.

Docstring saying one thing and code doing another isn't obvious.

tkelman replied:
> If the Dates section of the stdlib docs includes other docstrings for non-exported types, then this should be added.

So does it?

quinnj replied:
I added Time docs to the stdlib.

tkelman commented on base/dates/arithmetic.jl:
```diff
@@ -57,14 +72,16 @@ function (-)(dt::Date,z::Month)
+(+)(x::Time, y::TimePeriod) = return Time(Nanosecond(value(x) + tons(y)))
+(-)(x::Time, y::TimePeriod) = return Time(Nanosecond(value(x) - tons(y)))
+(+)(y::Period, x::TimeType) = x + y
+(-)(y::Period, x::TimeType) = x - y
```
are these magnitudes then?

quinnj replied:
Not sure I understand your comment. These are generic fallbacks that will end up dispatching to the definitions above. This is just for the switched order of operators.

tkelman replied:
It looks wrong that `y - x = x - y`

quinnj replied:
Ah, got it. Yes, these are magnitudes. Order doesn't matter.

tkelman replied:
worth a comment then, in case anyone else gets confused looking at this

tkelman replied:
How is this correct?
```
julia> Dates.Hour(24) - Dates.Time(Dates.Hour(1))
01:00:00
```

quinnj replied:
Because we already do
```julia
julia> Dates.Day(1) - Date(2016, 1, 1)
2015-12-31
```

tkelman replied:
wat? why allow that?

tkelman replied:
wat? why allow that?

quinnj replied:
I don't remember if it was an explicit decision to allow or not. We can open another issue to remove.

omus replied:
I'm in favor of removing this.

tkelman replied:
opened https://github.com/JuliaLang/julia/issues/20205

tkelman commented on base/dates/periods.jl:
```diff
@@ -16,18 +16,20 @@ for period in (:Year, :Month, :Week, :Day, :Hour, :Minute, :Second, :Millisecond
-    reference = period == :Week ? " For details see [`$accessor_str(::$typ_str)`](@ref)." : ""
+    typs = period in (:Microsecond, :Nanosecond) ? ["Time"] :
+           period in (:Hour, :Minute, :Second, :Millisecond) ? ["Time", "DateTime"] : ["Date","DateTime"]
+    reference = period == :Week ? " For details see [`$accessor_str(::Union{Date,DateTime})`](:func:`$accessor_str`)." : ""
```
old-style cross-reference format

quinnj replied:
What's the new style?

tkelman replied:
`(@ref)`, the way the line used to be

tkelman commented on base/dates/accessors.jl:
```diff
@@ -122,3 +129,14 @@ for parts in (["year", "month"], ["month", "day"], ["year", "month", "day"])
+            $($name)(t::Time) -> Int64
+
+         The $($name) of a `Time` as an `Int64`.
+         """ $func(t::Time)
```
spacing is off

quinnj replied:
How so? This follows the spacing of the block directly above it....

tkelman replied:
L134: 4 spaces
L135: 4 spaces
L136: 8 spaces
L137: 12 spaces
L138: NA
L139: 9 spaces <- should be 8
L140: 9 spaces <- should be 8
L141: 4 spaces

quinnj replied:
Ah, thanks. Sorry for missing this. Fixed.

quinnj committed:
Docs cleanup

quinnj commented:
Docs updated.

quinnj committed:
Fix whitespace

tkelman commented on base/dates/adjusters.jl:
```diff
@@ -199,7 +199,9 @@ provided `h, mi, s, ms, us` arguments, and will be adjusted until `f::Function`
-throwing an error (in the case that `f::Function` is never satisfied).
+throwing an error (in the case that `f::Function` is never satisfied). Note that the default step
+will adjust to allow for greater precision for the given arguments; i.e. if hour, minute, and second
+arguments are provided, the default step will be `Millisecond(1)` instead of `Second(1)`.
```
I'd either leave out the default in the docstring signature and say something like "the default value of `step` is the smaller of `Second(1)` or 1 unit of the next smaller precision than provided as input" in the description, or probably clearer, just write out the different signatures since square brackets isn't Julia syntax anyway
```
    Time(f::Function, h, mi=0; step=Second(1), negate=false, limit=10000) -> Time
    Time(f::Function, h, mi, s; step=Millisecond(1), negate=false, limit=10000) -> Time
    Time(f::Function, h, mi, s, ms; step=Microecond(1), negate=false, limit=10000) -> Time
    Time(f::Function, h, mi, s, ms, us; step=Nanosecond(1), negate=false, limit=10000) -> Time
```

tkelman replied:
was not addressed, docstring should not list a default value that isn't correct

tkelman commented on test/dates/ranges.jl:
```diff
@@ -503,3 +499,42 @@ let d = Dates.Day(1)
+@test a in dr
+
+@test all(x->sort(x) == (step(x) < zero(step(x)) ? reverse(x) : x),drs)
+@test all(x->step(x) < zero(step(x)) ? issorted(reverse(x)) : issorted(x),drs)
```
space after the comma in these `all` and `map` calls would look quite a bit better

tkelman replied:
was not addressed

quinnj replied:
Unfortunately, the entire dates module/files have inconsistent spacing. I can do a separate PR to do a bunch of these spacing issues all at once.

quinnj replied:
Though I'll probably wait until after the datetime-parsing PR so they don't have to rebase as much.

tkelman replied:
newly added lines aren't going to conflict with the other PR if it won't need to touch them

tkelman commented on test/dates/types.jl:
```diff
@@ -168,6 +192,12 @@ ms = Dates.Millisecond(1)
 @test Dates.Date(m,y) == Dates.Date(1,1,1)
-
-@test isfinite(Dates.Date)
-@test isfinite(Dates.DateTime)
```
these `isfinite` lines probably shouldn't be deleted

quinnj replied:
Added them back in.

quinnj committed:
Fix Time docs

quinnj committed:
Add isfinite tests for TimeTypes

tkelman commented:
- [ ] Should have a NEWS entry
- [ ] tests should use different values for fields instead of all ones https://github.com/JuliaLang/julia/pull/12274#discussion_r94787398
- [ ] commutative subtraction methods should be deleted https://github.com/JuliaLang/julia/issues/20205
- [ ] `Time` docstring signature should not say the default value of `step` is `Second(1)` since that's not always the default value https://github.com/JuliaLang/julia/pull/12274#discussion_r96987852
- [ ] formatting cleanup like https://github.com/JuliaLang/julia/pull/12274#discussion_r97016364

tkelman commented:
> I'd like to refactor a little bit to make the internal precision parameterised on a period (i.e. you could have a Second-precision Time, Millisecond-precision, etc.).

Correct me if I'm wrong, but that never happened? What changed?

quinnj commented:
At the time, I thought it'd be more useful to have a `Time` object that could hold different internal precisions, but after trying out an idea or two, it didn't actually seem as useful as the simple rule of always having it at `Nanosecond` resolution.

StefanKarpinski commented:
@tkelman, @quinnj: can we capture those todos in an issue?

quinnj commented:
They're all addressed in https://github.com/JuliaLang/julia/pull/20226
